src/client/dfuse/pil4dfs/int_dfs.c

```diff
@@ -5495,6 +5527,332 @@ register_handler(int sig, struct sigaction *old_handler)
 	}
 }
 
+static int
+allocate_aio_ctx(io_context_t ctx, int depth)
+{
+	int i;
+
+	pthread_mutex_lock(&lock_aio);
+	if (aio_ctx_count >= MAX_AIO_CTX)
+		goto err;
+
+	for (i = 0; i < MAX_AIO_CTX; i++) {
+		if (aio_ctx_list[i].ctx == 0) {
+			aio_ctx_list[i].ctx        = ctx;
+			aio_ctx_list[i].depth      = depth;
+			aio_ctx_list[i].on_dfs     = false;
+			aio_ctx_list[i].inited     = false;
+			aio_ctx_list[i].eq_ev_list = NULL;
+			aio_ctx_count++;
+			pthread_mutex_unlock(&lock_aio);
+			return 0;
+		}
+	}
+
+err:
+	pthread_mutex_unlock(&lock_aio);
+	DS_ERROR(EBUSY, "aio context list is full");
+	return EBUSY;
+}
+
+static int
+query_aio_ctx_depth(io_context_t ctx, int *idxp)
+{
+	int i, depth;
+
+	pthread_mutex_lock(&lock_aio);
+	for (i = 0; i < MAX_AIO_CTX; i++) {
+		if (aio_ctx_list[i].ctx == ctx) {
+			*idxp = i;
+			depth = aio_ctx_list[i].depth;
+			pthread_mutex_unlock(&lock_aio);
+			return depth;
+		}
+	}
+	pthread_mutex_unlock(&lock_aio);
+	return 0;
+}
+
+static void
+free_aio_ctx(io_context_t ctx)
+{
+	int i, io_depth, ctx_idx;
+
+	io_depth = query_aio_ctx_depth(ctx, &ctx_idx);
+	if (io_depth == 0)
+		return;
+
+	pthread_mutex_lock(&lock_aio);
+	if (aio_ctx_list[ctx_idx].inited) {
+		/* if eq initialized */
+		for (i = 0; i < aio_ctx_list[ctx_idx].depth; i++) {
+			daos_eq_destroy(aio_ctx_list[ctx_idx].eq_ev_list[i].eq, 0);
+		}
+		D_FREE(aio_ctx_list[ctx_idx].eq_ev_list);
+		aio_eq_count -= aio_ctx_list[ctx_idx].depth;
+	}
+	aio_ctx_list[ctx_idx].ctx = 0;
+	aio_ctx_count--;
+	pthread_mutex_unlock(&lock_aio);
+}
+
+int
+io_setup(int maxevents, io_context_t *ctxp)
+{
+	int rc;
+
+	if (next_io_setup == NULL) {
+		next_io_setup = dlsym(RTLD_NEXT, "io_setup");
+		D_ASSERT(next_io_setup != NULL);
+	}
+
+	rc = next_io_setup(maxevents, ctxp);
+	if (rc < 0)
+		return rc;
+
+	allocate_aio_ctx(*ctxp, maxevents);
+	/* we ignore the result value here since we are not sure whether DFS files will
+	 * be involved or not.
```

Comment:
I don't understand this comment. please clarify

Reply:
When applications call io_setup(), we do not know whether the files on DFS will be accessed or not. File descriptor will be passed in io_submit() later. Even allocate_aio_ctx() fails, we ignore it here and return the return code of next_io_setup(maxevents, ctxp).

Reply:
i still don't understand what this has to do with ignoring a failure on allocation..

Reply:
ok. A warning message could be printed although the failure on allocation should not change the return code of next_io_setup().